src/sdcard/mod.rs

```diff
@@ -5,176 +5,387 @@
+                Csd::V2(ref contents) => Ok(contents.card_capacity_blocks()),
+            }
+        })?;
+        Ok(BlockCount(num_blocks))
+    }
+
+    /// Return the usable size of this SD card in bytes.
+    fn num_bytes(&mut self) -> Result<u64, Error> {
+        self.with_chip_select(|s| {
+            let csd = s.read_csd()?;
+            debug!("CSD: {:?}", csd);
+            match csd {
+                Csd::V1(ref contents) => Ok(contents.card_capacity_bytes()),
+                Csd::V2(ref contents) => Ok(contents.card_capacity_bytes()),
+            }
+        })
+    }
+
+    /// Can this card erase single blocks?
+    pub fn erase_single_block_enabled(&mut self) -> Result<bool, Error> {
+        self.with_chip_select(|s| {
+            let csd = s.read_csd()?;
+            match csd {
+                Csd::V1(ref contents) => Ok(contents.erase_single_block_enabled()),
+                Csd::V2(ref contents) => Ok(contents.erase_single_block_enabled()),
+            }
+        })
+    }
+
+    /// Read the 'card specific data' block.
+    fn read_csd(&mut self) -> Result<Csd, Error> {
+        match self.card_type {
+            Some(CardType::SD1) => {
+                let mut csd = CsdV1::new();
+                if self.card_command(CMD9, 0)? != 0 {
+                    return Err(Error::RegisterReadError);
+                }
+                self.read_data(&mut csd.data)?;
+                Ok(Csd::V1(csd))
+            }
+            Some(CardType::SD2 | CardType::SDHC) => {
+                let mut csd = CsdV2::new();
+                if self.card_command(CMD9, 0)? != 0 {
+                    return Err(Error::RegisterReadError);
+                }
+                self.read_data(&mut csd.data)?;
+                Ok(Csd::V2(csd))
+            }
+            None => Err(Error::CardNotFound),
         }
     }
 
-    fn cs_high(&self) -> Result<(), Error> {
-        self.cs
-            .borrow_mut()
-            .set_high()
-            .map_err(|_| Error::GpioError)
+    /// Read an arbitrary number of bytes from the card. Always fills the
+    /// given buffer, so make sure it's the right size.
+    fn read_data(&mut self, buffer: &mut [u8]) -> Result<(), Error> {
+        // Get first non-FF byte.
+        let mut delay = Delay::new();
+        let status = loop {
+            let s = self.receive()?;
+            if s != 0xFF {
+                break s;
+            }
+            delay.delay(&mut self.delayer, Error::TimeoutReadBuffer)?;
+        };
+        if status != DATA_START_BLOCK {
+            return Err(Error::ReadError);
+        }
+
+        for b in buffer.iter_mut() {
+            *b = 0xFF;
+        }
+        self.spi.transfer(buffer).map_err(|_e| Error::Transport)?;
+
+        let mut crc = u16::from(self.receive()?);
+        crc <<= 8;
+        crc |= u16::from(self.receive()?);
+
+        let calc_crc = crc16(buffer);
+        if crc != calc_crc {
+            return Err(Error::CrcError(crc, calc_crc));
```

Comment:
Shouldn't crc checking be optional? Or what's the use case of the `require_crc` flag?

Reply:
I saw this and did wonder what was going on. Making CRC optional was added in #38.

Reply:
This is now changed to be `use_crc` and if it's false, we don't enable the CRC and we don't do any CRC calculations, sending or receiving. If it's true we enable the CRC with CMD59 and we do send/receive the CRC16s on all blocks.